### ISSUE TITLE
Life Time Analysis Execution Time

### ISSUE THREAD

**Issue by mrigger** [graalvm/sulong#18]
[Implementing a life time analysis](https://github.com/graalvm/sulong/blob/f9aef9d0aaf70eec0333ca06861fb32c8a92f62a/projects/com.oracle.truffle.llvm/src/com/oracle/truffle/llvm/parser/LLVMLifeTimeAnalysisVisitor.java) of virtual registers gave a huge speedup on many benchmarks. However, due to its naive implementation is also significantly increased the gate time (from 25 to 45 minutes on my machine).

**Comment by chrisseaton** [comment]
We have several people in the team who are experts in fast register allocation (which is what this problem is, isn't it?). I think Christian Wimmer wrote the fast linear scan allocator in C1 for example. Maybe talk to him?

**Comment by mrigger** [comment]
It was actually Christian Wimmer who suggested the algorithm for the life time analysis. I implemented it in the easiest way, while also often computing intermediate results redundantly. The problem is actually easier than register allocation, since we only need to assign null to dead slots. I will improve the code as soon as possible.

**Comment by chrisseaton** [comment]
Could you allocate these lists and queue outside of the loop, and let their capacity be re-used for each iteration?

https://github.com/graalvm/sulong/blob/f9aef9d0aaf70eec0333ca06861fb32c8a92f62a/projects/com.oracle.truffle.llvm/src/com/oracle/truffle/llvm/parser/LLVMLifeTimeAnalysisVisitor.java#L81-L84

**Comment by mrigger** [comment]
I tried this but it didn't influence performance. I profiled the application and discovered a low hanging fruit: By [caching the successors of a basic block](https://github.com/graalvm/sulong/pull/49) performance significantly improves. For the GCC test cases, execution takes 1,262 seconds with lifetime analysis enabled, 556 seconds when the analysis is disabled, and 638 seconds when cashing the successors.
